patch/patch.go

```diff
@@ -92,6 +95,76 @@ func WalkPatches(dir string, fn func(string) error) error {
 	return nil
 }
 
+const (
+	fromTimestampPrefix = "From "
+	fromAuthorPrefix    = "From: "
+	datePrefix          = "Date: "
+	subjectPrefix       = "Subject: "
+)
+
+// Patch is a parsed Git patch file.
+type Patch struct {
+	FromTimestamp string
+	FromAuthor    string
+	Date          string
+	Subject       string
+	Content       string
+}
+
+// Read reads and parses a patch from r.
+func Read(r io.Reader) (*Patch, error) {
+	var h Patch
+	var subject strings.Builder
+	var content strings.Builder
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+		line := scanner.Text()
+		if line == "---" || content.Len() != 0 {
+			// Header is done: read the rest of the patch into Content. Technically, "---" could
+			// occur inside the commit message, so we might be giving up early. But even "git
+			// format-patch" and "git am" don't round-trip "---" ("format-patch" doesn't escape it,
+			// "am" cuts off the message), so don't worry about it here.
+			content.WriteString(line)
+			content.WriteString("\n")
+		} else if strings.HasPrefix(line, fromTimestampPrefix) && h.FromTimestamp == "" {
+			h.FromTimestamp = line[len(fromTimestampPrefix):]
+		} else if strings.HasPrefix(line, fromAuthorPrefix) && h.FromAuthor == "" {
+			h.FromAuthor = line[len(fromAuthorPrefix):]
+		} else if strings.HasPrefix(line, datePrefix) && h.Date == "" {
+			h.Date = line[len(datePrefix):]
+		} else if strings.HasPrefix(line, subjectPrefix) && subject.Len() == 0 {
+			subject.WriteString(line[len(subjectPrefix):])
+			subject.WriteString("\n")
+		} else {
+			subject.WriteString(line)
+			subject.WriteString("\n")
+		}
+	}
+	h.Subject = subject.String()
+	h.Content = content.String()
```

Comment:
If it's ever a problem to store the patch file in memory (so far, < 1MB), a few options:

* I was poking around some other potential patch tool improvements and found `git mailinfo`: https://git-scm.com/docs/git-mailinfo. It splits a patch file into multiple files plus stdout that we could re-assemble into the final file.
* We can do something similar ourselves, in-process.